### PR TITLE
run workflow only if modified file matches the configured paths

### LINKED ISSUE #4854: Optimize build pipeline by taking affected paths into account
Currently, the build pipeline runs for any changes, even though some changes are known as not affecting the build results. For instance,

1. A change in documentation won't affect the build results of PHPUnit tests, static analysis or coding standards check.
2. A change in the configuration of a certain tool (e.g. PHPUnit) won't affect other build tools (e.g. PHPStan).
3. A change driver- or platform-specific code won't affect the build results in other drivers or platforms.

With that in mind, we can annotate the build jobs with `paths` and `paths-ignore` ([documentation](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths)).

### PR DESCRIPTION
**Pull request by marulitua**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | https://github.com/doctrine/dbal/issues/4854

#### Summary

static analysis:
  - src/
  - static-analysis/
  - tests/

continuous integration:
  - src/
  - tests/
  - ci/
  - composer.json
  - composer.lock

coding standards:
  - src/
  - bin/
  - tests/

### CODE REVIEW

File: .github/workflows/coding-standards.yml [hunk 1]
```diff
@@ -4,9 +4,17 @@ on:
   pull_request:
     branches:
       - "*.x"
+    paths:
+      - "src/**"
+      - "bin/**"
+      - "tests/**"
```
Comment (morozov) [inline .github/workflows/coding-standards.yml]: This should also take into account the coding standard configuration (`phpcs.xml`) and implementation (`composer.*`).

File: .github/workflows/coding-standards.yml [hunk 2]
```diff
@@ -4,9 +4,17 @@ on:
...
   push:
     branches:
       - "*.x"
+    paths:
+      - "src/**"
+      - "bin/**"
+      - "tests/**"
```
Comment (morozov) [inline .github/workflows/coding-standards.yml]: Would it be possible to get rid of the duplication of the paths for each workflow?

File: .github/workflows/static-analysis.yml [hunk 3]
```diff
@@ -4,9 +4,17 @@ on:
   pull_request:
     branches:
       - "*.x"
+    paths:
+      - "src/**"
+      - "static-analysis/**"
+      - "tests/**"
```
Comment (morozov) [inline .github/workflows/static-analysis.yml]: This should also take into account the configuration files of the static analysis tools.

**Review by morozov — COMMENTED** [review]
A good way to validate the completeness of the paths might be to use `strace`, run the jobs locally (each tool once) and see what files they actually use or even attempt to use. For instance:
```
$ strace cat test.php 2>&1 | grep openat
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "test.php", O_RDONLY)  = 3
```
Then disregard all the paths that don't belong to the project. Here, we can see that `cat test.php` depends on `test.php`.

**Comment by marulitua** [comment]
@morozov Can you check on my update

File: .github/workflows/coding-standards.yml (hunk 2 above)
Comment (morozov) [reply .github/workflows/coding-standards.yml]: @marulitua please do not resolve the conversations that you didn't start. What's the resolution on this one?

File: .github/workflows/static-analysis.yml [hunk 4]
```diff
@@ -4,9 +4,23 @@ on:
...
+      - "composer.*"
+      - "phpstan.neon.dist"
+      - "psalm*"
```
Comment (morozov) [inline .github/workflows/static-analysis.yml]: Could you be more specific here? Probably, `psalm-strict.xml`, `psalm.xml.dist`? This dangling `*` which isn't part of an extension bothers me.

File: .github/workflows/coding-standards.yml [hunk 5]
```diff
@@ -4,9 +4,21 @@ on:
   pull_request:
     branches:
       - "*.x"
+    paths:
+      - "src/**"
+      - "bin/**"
```
Comment (morozov) [inline .github/workflows/coding-standards.yml]: Please sort alphabetically.

**Review by morozov — COMMENTED** [review]
Content-wise looks good. Please retarget against `2.13.x` since it's relevant there as well. You'll have to replace `src/` with `lib/` there.

**Comment by morozov** [comment]
Hmm… it looks like required jobs remain required even if they are skipped. Now I see that in [Debezium](https://github.com/debezium/debezium) where I spotted this approach, they don't mark any of the jobs as required. @greg0ire do you think it will be safe to leave the status check to the maintainer's discretion? 

Another thing is that the AppVeyor build still runs unconditionally. It would be a bummer not to optimize it as well since it's one of the slowest jobs (especially, in the older branches like `2.13.x`).

**Comment by greg0ire** [comment]
> @greg0ire do you think it will be safe to leave the status check to the maintainer's discretion?

If we have to choose between both features I think it's best to pick the `paths` one. I'm sure Github will come back with a better solution in the future. It would be great to report this though.

**Comment by morozov** [comment]
> It would be great to report this though.

Agree. Do you know where this could be reported?

**Comment by morozov** [comment]
It should be possible to apply this to AppVeyor as well ([reference](https://www.appveyor.com/docs/how-to/filtering-commits/#include-commits-2)).

**Comment by greg0ire** [comment]
> Agree. Do you know where this could be reported?

Well, there is [this feedback form](https://support.github.com/contact/feedback), and otherwise you have https://github.community/ which will provide workarounds . Neither will feel like a great way of reporting this I'm afraid.

**Comment by morozov** [comment]
Submitted to both ([ref](https://github.community/t/configuring-paths-ignore-paths-on-a-required-workflow-doesnt-work-as-expected/209233)).

**Comment by stof** [comment]
this misses one path: the workflow file itself.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
I believe we can tackle the AppVeyor part later. Thanks, @marulitua!

**Pull request merged**